contracts/masset/Masset.sol

```diff
@@ -510,20 +570,20 @@ contract Masset is
         internal
         returns (uint256 massetRedeemed)
     {
+        console.log("\n~~ REDEEM ~~");
         require(_recipient != address(0), "Must be a valid recipient");
         uint256 bAssetCount = _bAssetQuantities.length;
         require(bAssetCount > 0 && bAssetCount == _bAssets.length, "Input array mismatch");
 
-        // Get high level basket info
-        Basket memory basket = basketManager.getBasket();
-
         // Prepare relevant data
-        ForgePropsMulti memory props = basketManager.prepareForgeBassets(_bAssets, _bAssetQuantities, false);
+        RedeemProps memory props = basketManager.prepareRedeemBassets(_bAssets);
         if(!props.isValid) return 0;
 
+        Cache memory cache = _getCacheDetails();
+
         // Validate redemption
         (bool redemptionValid, string memory reason, bool applyFee) =
-            forgeValidator.validateRedemption(basket.failed, totalSupply(), basket.bassets, props.indexes, _bAssetQuantities);
+            forgeValidator.validateRedemption(false, cache.vaultBalanceSum, props.allBassets, props.indexes, _bAssetQuantities);
```

Comment:
It seems like some of the validation now happens earlier (in `prepareRedeemBassets`). For instance, if `basket.failed` holds the following precondition would fail: https://github.com/mstable/mStable-contracts/blob/gas-reductions/contracts/masset/BasketManager.sol#L105. Previously, it seems like this would fail and it would be the validator's responsibility: https://github.com/mstable/mStable-contracts/blob/gas-reductions/contracts/masset/forge-validator/ForgeValidator.sol#L227. Is this correct and the intended behavior?

Reply:
Yes that is intentional - this single piece of validation was moved to the `prepareRedeemBassets`. The goal was to reduce the number of calls to BasketManager and consequently the quantity of data being returned.

Reply:
I see. Thanks! Would it make sense to update the validator to eliminate the code that is probably dead now?

Reply:
It would, but I weighed up the effort of doing another contract deploy and upgrade against simply setting this flag to false, and decided to just set the flag here.

Reply:
Ok, that makes sense. It might be good to keep track of such things in case you're redeploying contracts anyways.